src/operator/nn/pool.cuh

```diff
@@ -0,0 +1,826 @@
+  // index is the input image index in NCDHW
+  CUDA_KERNEL_LOOP(index, nthreads) {
+	  // find out the local index
+	  // find out the local offset
+	  const int w = index % width + pad_w;
+	  const int h = (index / width) % height + pad_h;
+    const int d = (index / width / height) % depth + pad_d;
+	  const int c = (index / width / height / depth) % channels;
+	  const int n = index / width / height / depth / channels;
+    const int pdstart = (d < kernel_d) ? 0 : (d - kernel_d) / stride_d + 1;
+    const int pdend = min(d / stride_d + 1, pooled_depth);
+	  const int phstart = (h < kernel_h) ? 0 : (h - kernel_h) / stride_h + 1;
+	  const int phend = min(h / stride_h + 1, pooled_height);
+	  const int pwstart = (w < kernel_w) ? 0 : (w - kernel_w) / stride_w + 1;
+	  const int pwend = min(w / stride_w + 1, pooled_width);
+	  DType gradient = 0;
+	  const DType* out_grad_slice =
+      out_grad + (n * channels + c) * pooled_depth * pooled_height * pooled_width;
+    for (int pd = pdstart; pd < pdend; ++pd) {
+      for (int ph = phstart; ph < phend; ++ph) {
+        for (int pw = pwstart; pw < pwend; ++pw) {
+          // figure out the pooling size
+          int dstart = pd * stride_d - pad_d;
+          int hstart = ph * stride_h - pad_h;
+          int wstart = pw * stride_w - pad_w;
+          int dend = min(dstart + kernel_d, depth + pad_d);
+          int hend = min(hstart + kernel_h, height + pad_h);
+          int wend = min(wstart + kernel_w, width + pad_w);
+          int pool_size = (isAvg? (dend - dstart) * (hend - hstart) * (wend - wstart) : 1);
+          gradient += out_grad_slice[(pd * pooled_height + ph) * pooled_width + pw] / pool_size;
+        }
+      }
+    }
+    // if req=kWriteTo, in_grad has already been assigned zero values in unpool()
+    // use "+=" here instead of "=" to accommodate when req=kAddTo
+	  in_grad[index] += gradient;
+  }
+}
+
+/*!
+ * \brief This function serves as an interface for 1/2/3-D pooling operations.
+ * \param s context stream defining the device in use is gpu
+ * \param in_data pointer of the input tensor data in the format of NCW, NCHW, or NCDHW
+ * \param ishape input tensor shape
+ * \param oshape output tensor shape
+ * \param kernel kernel shape
+ * \param pad pad shape
+ * \param stride stride shape
+ * \param pool_type supported pooling type: max, avg, sum
+ * \param req_type operator request type: kNullOp, kNullWriteInplace, kNullWriteTo, kNullAddTo
+ * \param out_data pointer of the output tensor data in the format of NCW, NCHW, or NCDHW
+ */
+template<typename DType>
+inline void pool(mshadow::Stream<gpu>* s, const DType* in_data, const TShape& ishape,
+                 const TShape& oshape, const TShape& kernel, const TShape& pad,
+                 const TShape& stride, const int pool_type, OpReqType req_type,
+                 DType* out_data) {
+  using namespace mxnet_op;
+  if (kernel.ndim() == 1) {
```

Comment:
this is a little to much copy pasting.
how about change unpool_sum_?d_gpu_kernel to unpool_sum_gpu_kernel(..., Shape<?>)
and add int ndim to pool's template argument